File: aat-android/src/main/java/ch/bailu/aat/dispatcher/TrackerSource.java
package ch.bailu.aat.dispatcher;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat.util.OldAppBroadcaster;
import ch.bailu.aat_lib.dispatcher.AppBroadcaster;
import ch.bailu.aat_lib.dispatcher.ContentSource;
import ch.bailu.aat_lib.gpx.GpxInformation;
import ch.bailu.aat_lib.gpx.InfoID;

public class TrackerSource extends ContentSource {

    private final ServiceContext scontext;

    public TrackerSource (ServiceContext sc) {
        scontext = sc;

    }

    private final BroadcastReceiver onTrackChanged = new BroadcastReceiver () {
        @Override
        public void onReceive(Context context, Intent intent) {
            sendUpdate(InfoID.TRACKER, scontext.getTrackerService().getLoggerInformation());
        }

    };


    @Override
    public void requestUpdate() {
        sendUpdate(InfoID.TRACKER, scontext.getTrackerService().getLoggerInformation());
    }


    @Override
    public void onPause() {
        scontext.getContext().unregisterReceiver(onTrackChanged);
    }


    @Override
    public void onResume() {
        OldAppBroadcaster.register(scontext.getContext(), onTrackChanged, AppBroadcaster.TRACKER);
    }

    @Override
    public int getIID() {
        return InfoID.TRACKER;
    }

    @Override
    public GpxInformation getInfo() {
        return scontext.getTrackerService().getLoggerInformation();
    }


}

File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/source/TrackerSource.kt
package ch.bailu.aat_lib.dispatcher.source

import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.broadcaster.Broadcaster
import ch.bailu.aat_lib.dispatcher.SourceInterface
import ch.bailu.aat_lib.dispatcher.TargetInterface
import ch.bailu.aat_lib.dispatcher.usage.UsageTrackerInterface
import ch.bailu.aat_lib.gpx.information.GpxInformation
import ch.bailu.aat_lib.gpx.information.InfoID
import ch.bailu.aat_lib.service.ServicesInterface

class TrackerSource(
    private val services: ServicesInterface,
    private val broadcaster: Broadcaster,
    usageTracker: UsageTrackerInterface
) :
    SourceInterface
{
    private var target = TargetInterface.NULL

    init {
        usageTracker.observe {
            requestUpdate()
        }
    }

    override fun setTarget(target: TargetInterface) {
        this.target = target
    }

    private val onTrackChanged = BroadcastReceiver { requestUpdate() }

    override fun requestUpdate() {
        target.onContentUpdated(InfoID.TRACKER, services.getTrackerService().info)
    }

    override fun onPauseWithService() {
        broadcaster.unregister(onTrackChanged)
    }

    override fun onDestroy() {}

    override fun onResumeWithService() {
        broadcaster.register(AppBroadcaster.TRACKER, onTrackChanged)
    }

    override fun getIID(): Int {
        return InfoID.TRACKER
    }

    override fun getInfo(): GpxInformation {
        return services.getTrackerService().info
    }
}
